google/cloud/spanner/mutations_test.cc

```diff
@@ -0,0 +1,204 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "google/cloud/spanner/mutations.h"
+#include <google/protobuf/text_format.h>
+#include <google/protobuf/util/message_differencer.h>
+#include <gmock/gmock.h>
+
+namespace google {
+namespace cloud {
+namespace spanner {
+inline namespace SPANNER_CLIENT_NS {
+namespace {
+
+using ::testing::HasSubstr;
+
+TEST(MutationsTest, Default) {
+  Mutation actual;
+  EXPECT_EQ(actual, actual);
+}
+
+TEST(MutationsTest, PrintTo) {
+  Mutation insert = MakeInsertMutation(std::string("test-string"));
+  std::ostringstream os;
+  PrintTo(insert, &os);
+  auto actual = std::move(os).str();
+  EXPECT_THAT(actual, HasSubstr("test-string"));
+  EXPECT_THAT(actual, HasSubstr("Mutation={"));
+}
+
+TEST(MutationsTest, InsertSimple) {
+  Mutation empty;
+  Mutation insert =
+      MakeInsertMutation(std::string("foo"), std::string("bar"), true);
+  EXPECT_EQ(insert, insert);
+  EXPECT_NE(insert, empty);
+
+  auto actual = std::move(insert).as_proto();
+  google::spanner::v1::Mutation expected;
+  ASSERT_TRUE(google::protobuf::TextFormat::ParseFromString(R"""(
+              insert: {
+                values {
+                  values {
+                    string_value: "foo"
+                  }
+                  values {
+                    string_value: "bar"
+                  }
+                  values {
+                    bool_value: true
+                  }
+                }
+              }
+              )""",
+                                                            &expected));
+  std::string delta;
+  google::protobuf::util::MessageDifferencer differencer;
+  differencer.ReportDifferencesToString(&delta);
+  EXPECT_TRUE(differencer.Compare(actual, expected)) << delta;
+}
+
+TEST(MutationsTest, InsertComplex) {
+  Mutation empty;
+  auto builder = InsertMutationBuilder({"col1", "col2", "col3"})
+                     .AddRow(std::int64_t(42), std::string("foo"), false)
```

Comment:
(related to a previous comment)

If we switched to world where the argument type is `Row<Ts...>` and callers used `MakeRow()`, they would not need to go out of their way to cast `42` as `std::int64_t(42)`, because `MakeRow` would do that implicitly.

FWIW

Reply:
I changed the test to show that `42` also works. Under the hood all I am doing is `internal::ToProto(Value(parameter))` so any conversion that works for `Value` should also work here.